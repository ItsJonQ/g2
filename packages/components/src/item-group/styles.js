import { css, ui } from '@wp-g2/styles';

// @todo: Maybe abstract to a dedicated UnstyledButton component.
export const unstyledButton = css`
	appearance: none;
	background: none;
	border: 1px solid transparent;
	cursor: pointer;
	text-align: left;

	&:hover {
		color: ${ui.get('colorAdmin')};
	}

	&:focus {
		background-color: transparent;
		border-color: ${ui.get('colorAdmin')};
		color: ${ui.get('colorAdmin')};
		outline: 3px solid transparent;
	}
`;

export const item = css`
	display: block;
	width: 100%;
`;

export const bordered = css`
	${ui.border.all}
`;

export const separated = css`
	> *:not(marquee) {
		${ui.border.bottom}
	}

	> *:last-child:not(:focus) {
		border-bottom-color: transparent;
	}
`;

const borderRadius = ui.get('controlBorderRadius');

export const spacedAround = css`
	${ui.borderRadius(borderRadius)}
`;

export const rounded = css`
	${ui.borderRadius(borderRadius)}

	> *:first-child {
		border-top-left-radius: ${borderRadius};
		border-top-right-radius: ${borderRadius};
	}

	> *:last-child {
		border-bottom-left-radius: ${borderRadius};
		border-bottom-right-radius: ${borderRadius};
	}
`;

const baseFontHeight = `calc(${ui.get('fontSize')} * ${ui.get(
	'fontLineHeightBase',
)})`;

/*
 * Math:
 * - Use the desired height as the base value
 * - Subtract the computed height of (default) text
 * - Subtract the effects of border
 * - Divide the calculated number by 2, in order to get an individual top/bottom padding
 */
const paddingY = `calc((${ui.get(
	'controlHeight',
)} - ${baseFontHeight} - 2px) / 2)`;
const paddingYSmall = `calc((${ui.get(
	'controlHeightSmall',
)} - ${baseFontHeight} - 2px) / 2)`;
const paddingYLarge = `calc((${ui.get(
	'controlHeightLarge',
)} - ${baseFontHeight} - 2px) / 2)`;

export const itemSizes = {
	small: css`
		${ui.padding.y(paddingYSmall)}
		${ui.padding.x(ui.get('controlPaddingXSmall'))}
	`,
	medium: css`
		${ui.padding.y(paddingY)}
		${ui.padding.x(ui.get('controlPaddingX'))}
	`,
	large: css`
		${ui.padding.y(paddingYLarge)}
		${ui.padding.x(ui.get('controlPaddingXLarge'))}
	`,
};
