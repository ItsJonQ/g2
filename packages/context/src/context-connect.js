import { uniq } from 'lodash';
import React, { forwardRef } from 'react';

import { CONNECT_STATIC_NAMESPACE } from './constants';

/**
 * Forwards ref (React.ForwardRef) and "Connects" (or registers) a component
 * within the Context system under a specified namespace.
 *
 * This is an (experimental) evolution of the initial connect() HOC.
 * The hope is that we can improve render performance by removing functional
 * component wrappers.
 *
 * @template P
 * @param {import('react').ForwardRefRenderFunction<import('@wp-g2/create-styles').ElementTypeFromViewOwnProps<P>, P>} Component The component to register into the Context system.
 * @param {Array<string>|string} namespace The namespace to register the component under.
 * @param {object} options
 * @param {boolean} [options.memo=true]
 * @return {import('@wp-g2/create-styles').PolymorphicComponent<import('@wp-g2/create-styles').ElementTypeFromViewOwnProps<P>, P>}
 */
export function contextConnect(Component, namespace, options = {}) {
	const { memo = true } = options;

	let WrappedComponent = forwardRef(Component);
	if (memo) {
		// @ts-ignore
		WrappedComponent = React.memo(WrappedComponent);
	}

	const displayName = Array.isArray(namespace)
		? namespace[0]
		: namespace || WrappedComponent.name;

	let mergedNamespace = WrappedComponent[CONNECT_STATIC_NAMESPACE] || [
		displayName,
	];

	/**
	 * Consolidate (merge) namespaces before attaching it to the WrappedComponent.
	 */
	if (Array.isArray(namespace)) {
		mergedNamespace = [...mergedNamespace, ...namespace];
	}
	if (typeof namespace === 'string') {
		mergedNamespace = [...mergedNamespace, namespace];
	}

	WrappedComponent.displayName = displayName;
	WrappedComponent[CONNECT_STATIC_NAMESPACE] = uniq(mergedNamespace);

	// @ts-ignore
	return WrappedComponent;
}

/**
 * Attempts to retrieve the connected namespace from a component.
 *
 * @param {import('react').ReactChild | undefined | {}} Component The component to retrieve a namespace from.
 * @returns {Array<string>} The connected namespaces.
 */
export function getConnectNamespace(Component) {
	if (!Component) return [];

	let namespaces = [];

	if (Component[CONNECT_STATIC_NAMESPACE]) {
		namespaces = Component[CONNECT_STATIC_NAMESPACE];
	}

	// @ts-ignore
	if (Component.type && Component.type[CONNECT_STATIC_NAMESPACE]) {
		// @ts-ignore
		namespaces = Component.type[CONNECT_STATIC_NAMESPACE];
	}

	return namespaces;
}

/**
 * Checks to see if a component is connected within the Context system.
 *
 * @param {import('react').ReactNode} Component The component to retrieve a namespace from.
 * @param {Array<string>|string} match The namespace to check.
 * @returns {boolean} The result.
 */
export function hasConnectNamespace(Component, match) {
	if (!Component) return false;

	if (typeof match === 'string') {
		return getConnectNamespace(Component).includes(match);
	}
	if (Array.isArray(match)) {
		return match.some((result) =>
			getConnectNamespace(Component).includes(result),
		);
	}

	return false;
}

export const getNamespace = getConnectNamespace;
export const hasNamespace = hasConnectNamespace;
