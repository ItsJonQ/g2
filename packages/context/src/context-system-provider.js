import { deepEqual, deepMerge, useIsomorphicLayoutEffect } from '@wp-g2/utils';
import { isNil } from 'lodash';
import React, { createContext, useContext, useRef, useState } from 'react';

export const ComponentsContext = createContext({});
export const useComponentsContext = () => useContext(ComponentsContext);

/**
 * Consolidates incoming ContextSystem values with a (potential) parent ContextSystem value.
 *
 * @param {object} props
 * @param {object} props.value
 * @return {object}
 */
function useContextSystemBridge({ value }) {
	if (isNil(value)) {
		// @ts-ignore
		value = {};
	}
	const parentContext = useComponentsContext();
	const parentContextRef = useRef(parentContext);
	const valueRef = useRef(deepMerge(parentContext, value));

	const [config, setConfig] = useState(valueRef.current);

	useIsomorphicLayoutEffect(() => {
		let hasChange = false;

		if (!deepEqual(value, valueRef.current)) {
			valueRef.current = value;
			hasChange = true;
		}

		if (!deepEqual(parentContext, parentContextRef.current)) {
			valueRef.current = deepMerge(parentContext, valueRef.current);
			parentContextRef.current = parentContext;
			hasChange = true;
		}

		if (hasChange) {
			setConfig((prev) => ({ ...prev, ...valueRef.current }));
		}
	}, [value, parentContext]);

	return config;
}

/**
 * A Provider component that can modify props for connected components within
 * the Context system.
 *
 * @example
 * ```jsx
 * <ContextSystemProvider value={{ Button: { size: 'small' }}}>
 *   <Button>...</Button>
 * </ContextSystemProvider>
 * ```
 *
 * @template {Record<string, any>} T
 * @param {object} options
 * @param {import('react').ReactNode} options.children Children to render.
 * @param {T} options.value Props to render into connected components.
 * @returns {JSX.Element} A Provider wrapped component.
 */
const BaseContextSystemProvider = ({ children, value }) => {
	const contextValue = useContextSystemBridge({ value });

	return (
		<ComponentsContext.Provider value={contextValue}>
			{children}
		</ComponentsContext.Provider>
	);
};

export const ContextSystemProvider = React.memo(BaseContextSystemProvider);
