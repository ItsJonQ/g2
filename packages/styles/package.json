{
    "name": "@wp-g2/styles",
    "version": "0.0.47",
    "description": "Style system for G2 Components.",
    "main": "dist/cjs/index.js",
    "module": "dist/es/index.js",
    "sideEffects": false,
    "types": "types",
    "scripts": {
        "clean": "rm -rf dist",
        "build:es": "BUILD_FORMAT=es zero build -d dist/es",
        "build:cjs": "BUILD_FORMAT=cjs zero build -d dist/cjs",
        "build": "npm run build:cjs && npm run build:es -- --no-clean",
        "test": "echo \"Error: no test specified\" && exit 1"
    },
    "author": "Jon Quach <hello@jonquach.com> (https://jonquach.com)",
    "license": "MIT",
    "dependencies": {
        "@emotion/core": "^10.0.28",
        "@emotion/is-prop-valid": "^0.8.8",
        "@wp-g2/utils": "^0.0.47",
        "create-emotion": "^10.0.27",
        "cssjanus": "^1.3.2",
        "dash-get": "^1.0.2",
        "emotion": "^10.0.27",
        "emotion-theming": "^10.0.27",
        "styled-griddie": "^0.1.3",
        "stylis-plugin-css-variables": "^1.0.4",
        "stylis-plugin-rtl": "^1.0.0"
    },
    "peerDependencies": {
        "react": "^16.8.0",
        "react-dom": "^16.8.0"
    }
}
