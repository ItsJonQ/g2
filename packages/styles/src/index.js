export * from './core';
export { compiler } from './system';
export { cache, css, cx, injectGlobal, keyframes } from './style-system';

export * from './styled';
export * from './hooks';
export * from './mixins';
export * from './namespaces';
export {
	config,
	createTheme,
	getIsReducedMotion,
	SUPPORTED_COLORS,
} from './theme';

export * from './presets/ui';

export * from './components';
