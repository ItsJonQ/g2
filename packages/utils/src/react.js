import { Children, isValidElement } from 'react';

/**
 * Merges React `ref` together.
 */
export { default as mergeRefs } from 'react-merge-refs';

/**
 * Copies non-react specific statics from a child component to a parent component.
 * Similar to Object.assign, but with React static keywords prevented from being overridden.
 */
export { default as hoistNonReactStatics } from 'hoist-non-react-statics';

/**
 * Gets a collection of available children elements from a React component's children prop.
 *
 * @param {import('react').ReactNode} children
 *
 * @return {import('react').ReactNodeArray} An array of available children.
 */
export function getValidChildren(children) {
	if (typeof children === 'string') return [children];

	return Children.toArray(children).filter((child) => isValidElement(child));
}

/**
 * Gets the displayName of a React Component or element.
 *
 * @param {string | import('react').ComponentType} tagName
 *
 * @return {string} The display name of the Component / tagName.
 */
export function getDisplayName(tagName) {
	const displayName =
		typeof tagName === 'string'
			? tagName
			: tagName.displayName || tagName.name || 'Component';

	return displayName;
}

/**
 * Checks to see if a `children` prop is a render function prop.
 *
 * @param {*} children
 *
 * @return {boolean} True, if children is a render function prop.
 */
export function isRenderProp(children) {
	return typeof children === 'function';
}

/**
 * Handles the rendering of a React component's children prop, which
 * may be a render function prop.
 *
 * @param {*} children Children to render.
 * @param {*} props Props to pass into a (potential) children render function.
 *
 * @return {*} The rendered children.
 */
export function renderChildren(children, props = {}) {
	if (isRenderProp(children)) {
		// eslint-disable-next-line no-unused-vars
		const { children: _, ...rest } = props;
		return children(rest);
	}

	return children;
}
