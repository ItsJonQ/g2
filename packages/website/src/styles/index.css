body {
  padding-top: 0;
}
