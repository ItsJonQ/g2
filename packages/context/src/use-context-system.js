import { css, cx } from '@wp-g2/styles';
import { memoize } from '@wp-g2/utils';
import { kebabCase, uniq } from 'lodash';

import { CONNECTED_NAMESPACE } from './constants';
import { useComponentsContext } from './context-system-provider';
import { ns } from './utils';

/**
 * @template TProps
 * @typedef {TProps & { [CONNECTED_NAMESPACE]: boolean; className: string; children?: import('react').ReactNode }} ConnectedProps
 */

/**
 * Custom hook that derives registered props from the Context system.
 * These derived props are then consolidated with incoming component props.
 *
 * @template {{ className?: string }} P
 * @param {P} props Incoming props from the component.
 * @param {string} namespace The namespace to register and to derive context props from.
 * @return {ConnectedProps<P>}
 */
export function useContextSystem(props, namespace) {
	const contextSystemProps = useComponentsContext();
	const displayName = Array.isArray(namespace) ? namespace[0] : namespace;

	const contextProps = contextSystemProps?.[displayName] || {};

	/** @type {ConnectedProps<P>} */
	// @ts-ignore We fill in the missing properties below
	const finalComponentProps = {
		[CONNECTED_NAMESPACE]: true,
	};

	const nextNs = ns(displayName);
	for (const k in nextNs) {
		if (typeof nextNs[k] === 'string') {
			finalComponentProps[k] = nextNs[k];
		}
	}

	const {
		_overrides: overrideProps,
		css: contextCSS,
		...otherContextProps
	} = contextProps;

	const initialMergedProps = Object.entries(otherContextProps).length
		? Object.assign({}, otherContextProps, props)
		: props;

	const classes = cx(
		// Resolve custom CSS from ContextSystemProvider
		contextCSS && css(contextCSS),
		initialMergedProps.css && css(initialMergedProps.css),
		memoizedGetStyledClassNameFromKey(displayName),
		props.className,
	);

	// Provides the ability to customize the render of the component.
	const rendered =
		typeof initialMergedProps.renderChildren === 'function'
			? initialMergedProps.renderChildren(initialMergedProps)
			: initialMergedProps.children;

	for (const k in initialMergedProps) {
		/**
		 * Omitting CSS prop.
		 */
		if (k !== 'css') {
			finalComponentProps[k] = initialMergedProps[k];
		}
	}

	for (const k in overrideProps) {
		finalComponentProps[k] = overrideProps[k];
	}

	finalComponentProps.children = rendered;
	finalComponentProps.className = classes;

	return finalComponentProps;
}

/**
 * Generates the connected component CSS className based on the namespace.
 *
 * @param {string} displayName The name of the connected component.
 * @returns {string} The generated CSS className.
 */
function getStyledClassName(displayName) {
	if (!displayName || typeof displayName !== 'string') return '';

	const kebab = kebabCase(displayName);
	return `components-${kebab} wp-components-${kebab}`;
}

/**
 * Generates the connected component CSS className based on the namespace.
 *
 * @param {string} key The name of the connected component.
 * @returns {string} The generated CSS className.
 */
function getStyledClassNameFromKey(key) {
	if (!key) return '';

	if (Array.isArray(key)) {
		return cx(uniq(key).map(getStyledClassName));
	}
	if (typeof key === 'string') {
		return getStyledClassName(key);
	}

	return '';
}

const memoizedGetStyledClassNameFromKey = memoize(getStyledClassNameFromKey);
