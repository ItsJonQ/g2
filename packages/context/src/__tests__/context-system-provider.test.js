import { render } from '@testing-library/react';
import { View } from '@wp-g2/styles';
import React from 'react';

import { contextConnect } from '../context-connect';
import { ContextSystemProvider } from '../context-system-provider';
import { useContextSystem } from '../use-context-system';

describe('props', () => {
	test('should render correctly', () => {
		const Olaf = (props, ref) => (
			<View {...useContextSystem(props, 'Olaf')} ref={ref} />
		);
		const ConnectedOlaf = contextConnect(Olaf, 'Olaf');
		const { container } = render(
			<ContextSystemProvider>
				<ConnectedOlaf />
			</ContextSystemProvider>,
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render context props', () => {
		const Olaf = (props, ref) => {
			const { quote, ...otherProps } = useContextSystem(props, 'Olaf');
			return (
				<View {...otherProps} ref={ref}>
					{quote}
				</View>
			);
		};

		const ConnectedOlaf = contextConnect(Olaf, 'Olaf');

		const contextValue = {
			Olaf: {
				quote: 'Warm Hugs',
			},
		};

		const { container } = render(
			<ContextSystemProvider value={contextValue}>
				<ConnectedOlaf />
			</ContextSystemProvider>,
		);

		expect(container.firstChild).toMatchSnapshot();
		expect(container.firstChild.innerHTML).toContain('Warm Hugs');
	});

	test('should render css props', () => {
		const Olaf = (props, ref) => {
			const { quote, ...otherProps } = useContextSystem(props, 'Olaf');
			return (
				<View {...otherProps} ref={ref}>
					{quote}
				</View>
			);
		};

		const ConnectedOlaf = contextConnect(Olaf, 'Olaf');

		const contextValue = {
			Olaf: {
				css: [{ background: 'white' }, `padding: 20px;`],
			},
		};

		const { container } = render(
			<>
				<ContextSystemProvider value={contextValue}>
					<ConnectedOlaf className="olaf" />
				</ContextSystemProvider>
				<ConnectedOlaf className="olaf" />
			</>,
		);

		const [first, second] = [...container.querySelectorAll('.olaf')];

		expect(container).toMatchSnapshot();

		const firstCSS = first.classList.toString();
		const secondCSS = second.classList.toString();

		expect(first).toHaveStyle(`background: white`);
		expect(first).toHaveStyle(`padding: 20px`);

		expect(second).not.toHaveStyle(`background: white`);
		expect(second).not.toHaveStyle(`padding: 20px`);

		expect(firstCSS).not.toEqual(secondCSS);
	});

	test('should render css props + provided css', () => {
		const Olaf = (props, ref) => {
			const { quote, ...otherProps } = useContextSystem(props, 'Olaf');
			return (
				<View {...otherProps} ref={ref}>
					{quote}
				</View>
			);
		};

		const ConnectedOlaf = contextConnect(Olaf, 'Olaf');

		const contextValue = {
			Olaf: {
				css: [{ background: 'white' }, { border: '2px solid blue' }],
			},
		};

		const { container } = render(
			<>
				<ContextSystemProvider value={contextValue}>
					<ConnectedOlaf
						className="olaf"
						css={`
							font-weight: bold;
						`}
					/>
				</ContextSystemProvider>
			</>,
		);

		expect(container).toMatchSnapshot();

		const el = container.querySelector('.olaf');

		expect(el).toHaveStyle(`border: 2px solid blue`);
		expect(el).toHaveStyle(`background: white`);
		expect(el).toHaveStyle(`font-weight: bold`);
	});

	test('should render _override props', () => {
		const Olaf = (props, ref) => {
			const { quote, ...otherProps } = useContextSystem(props, 'Olaf');
			return (
				<View {...otherProps} ref={ref}>
					{quote}
				</View>
			);
		};

		const ConnectedOlaf = contextConnect(Olaf, 'Olaf');

		const contextValue = {
			Olaf: {
				_overrides: {
					quote: 'Warm Hugs!',
				},
			},
		};

		const { container } = render(
			<>
				<ContextSystemProvider value={contextValue}>
					<ConnectedOlaf className="olaf" quote="Hello" />
				</ContextSystemProvider>
			</>,
		);

		expect(container.firstChild).toMatchSnapshot();

		const el = container.querySelector('.olaf');

		expect(el.innerHTML).toContain('Warm Hugs!');
		expect(el.innerHTML).not.toContain('Hello');
	});
});
