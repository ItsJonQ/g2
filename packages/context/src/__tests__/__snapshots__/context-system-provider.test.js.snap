// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props should render _override props 1`] = `
<div
  class="components-olaf wp-components-olaf olaf css-1nqjm19"
  data-g2-c16t="true"
  data-g2-component="Olaf"
>
  Warm Hugs!
</div>
`;

exports[`props should render context props 1`] = `
<div
  class="components-olaf wp-components-olaf css-1nqjm19"
  data-g2-c16t="true"
  data-g2-component="Olaf"
>
  Warm Hugs
</div>
`;

exports[`props should render correctly 1`] = `
<div
  class="components-olaf wp-components-olaf css-1nqjm19"
  data-g2-c16t="true"
  data-g2-component="Olaf"
/>
`;

exports[`props should render css props + provided css 1`] = `
<div>
  <div
    class="components-olaf wp-components-olaf olaf css-w4hnsq"
    data-g2-c16t="true"
    data-g2-component="Olaf"
  />
</div>
`;

exports[`props should render css props 1`] = `
<div>
  <div
    class="components-olaf wp-components-olaf olaf css-hs2bux"
    data-g2-c16t="true"
    data-g2-component="Olaf"
  />
  <div
    class="components-olaf wp-components-olaf olaf css-1nqjm19"
    data-g2-c16t="true"
    data-g2-component="Olaf"
  />
</div>
`;
